File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

#define MAX(first, second) ( ( first ) > ( second ) ? ( first ) : ( second ) )

int diameterOfBinaryTreeRec(struct TreeNode *root, int *longestPathFromRoot) {
    if (root == NULL) {
        *longestPathFromRoot = -1;
        return -1;
    }

    int longestFromLeft = -1;
    int longestFromRight = -1;

    int diameterLeft = diameterOfBinaryTreeRec(root->left, &longestFromLeft);
    int diameterRight = diameterOfBinaryTreeRec(root->right, &longestFromRight);
    int throughRoot = longestFromLeft + longestFromRight + 2;

    *longestPathFromRoot = 1 + MAX(longestFromLeft, longestFromRight);
    return MAX(throughRoot, MAX(diameterLeft, diameterRight));
}

int diameterOfBinaryTree(struct TreeNode* root) {
    int longestPathFromRoot = -1;
    return diameterOfBinaryTreeRec(root, &longestPathFromRoot);
}

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    private int DiameterOfBinaryTreeRec(TreeNode root, out int depth) {
        if (root == null) {
            depth = -1;
            return -1;
        }

        int leftDepth;
        int rightDepth;

        int leftDiameter = DiameterOfBinaryTreeRec(root.left, out leftDepth);
        int rightDiameter = DiameterOfBinaryTreeRec(root.right, out rightDepth);

        depth = 1 + Math.Max(leftDepth, rightDepth);
        return Math.Max(leftDiameter, Math.Max(rightDiameter, 2 + leftDepth + rightDepth));
    }

    public int DiameterOfBinaryTree(TreeNode root) {
        int diameter;
        return DiameterOfBinaryTreeRec(root, out diameter);
    }
}